components/abtesting/optimizelyXExperiment/src/index.js

```diff
@@ -1,22 +1,119 @@
 /* eslint handle-callback-err: "off" */
-import PropTypes from 'prop-types'
-
 import React, {Component} from 'react'
+import PropTypes from 'prop-types'
 import AbTestToggle from '@s-ui/abtesting-toggle'
+
+import ExperimentContext from './experiment-context'
 import OptimizelyX from './optimizely-x'
 
 class AbTestOptimizelyXExperiment extends Component {
-  state = {variation: null}
+  constructor(props) {
+    super(props)
+
+    this.defaultVariation = this._getDefaultVariation()
+    this.initialVariationId =
+      this._getForceVariationId() || this.defaultVariation
+    this.initialVariationName = this._getVariationNameFromVariationId(
+      this.initialVariationId
+    )
+
+    // no need to pass the special children prop to the context
+    const {children, ...propsWithoutChildren} = props
+
+    this.state = {
```

Comment:
In order to simplify and avoid duplicating calculating some info, you could avoid using state for fields that you could calculate once on the `render` and pass on the object.

For example, `isDefault` and `isVariation` could be calculated on the render using the `variationId` from the state.

This will help you to simplify the code and do some checks only in one place.

Reply:
I see, but I'm putting all that stuff in the state because of [this warning from the React docs](https://es.reactjs.org/docs/context.html#caveats), so the value is passed as a unique reference to the provider (the state itself) and not as a new object at every render.

Does it make any sense? Or am I missing something?

Reply:
Make totally sense. What about extracting a function the creation of the state in order to be used for the initialState and the update? At least, you will reuse the state logic and keep it in one place.

Reply:
Oh, *laik* 😄

Reply:
### ✅ Done

Tests keep passing in my local environment and it looks much better now, thanks!